### PR TITLE
ExclusiveTopology with ArrayOfVectorViews backend

### PR DESCRIPTION
**Pull request by KnutAM**
See discussion in #965, this PR introduces a cleaned-up version. 

Most important changes in this PR compared to master are:
* Storage from `Array{EntityNeighborhood{T}, N}` -> `ArrayOfVectorViews{T, N} <: AbstractArray{"View{Vector{T}}", N}`
* `getcells` dispatches removed (since these would now act on `Vector{<:BoundaryIndex}`). In any case, `getcells` normally returns `::AbstractCell`s, but in topology they return `::Int`s on mater, if any, `cellid` should be overloaded).
* `EntityNeighborhood` is gone

Compared to #965, the support for hashed indexing is removed and only `ExclusiveTopology` is implemented. Before finalizing, I would like some feedback from the topology side, and then I'll add

- [x] Separate tests for `ArrayOfVectorViews`
- [x] Documentation of `ArrayOfVectorViews`
- [x] Add warning for non-conforming grids, ref https://github.com/Ferrite-FEM/Ferrite.jl/issues/941#issuecomment-2162836723 (and for embedded grids)
- [x] Update docstrings for `ExclusiveTopology`
- [x] Changelog

Let me know if more details are required for the initial review. 

Note that in relation to #942, this PR explicitly errors if trying to construct `ExclusiveTopology` with grids containing embedded cells, thus avoiding silent bugs.

**Benchmarks**

```julia
using Ferrite, BenchmarkTools
function sum_indices(nbh::AbstractMatrix)
    v = 0
    for nbs in nbh
        for n in nbs
            v += n[1] + n[2]
        end
    end
    return v
end

function sum_indices(top::ExclusiveTopology)
    v  = sum_indices(top.vertex_vertex_neighbor)
    v += sum_indices(top.edge_edge_neighbor)
    v += sum_indices(top.face_face_neighbor)
    return v
end
function sum_skeleton_facet_neighbors(top, grid)
    skeleton = Ferrite.facetskeleton(top, grid)
    v = 0
    for facetidx in skeleton
        v += sum(n[1] + n[2] for n in getneighborhood(top, grid, facetidx); init = 0)
    end
    return v
end

grid = generate_grid(Hexahedron, (10, 10, 10));
top = ExclusiveTopology(grid); # Compilation

@btime sum_indices($top); 
# Master: 55.700 μs (0 allocations: 0 bytes)
#     PR: 36.000 μs (0 allocations: 0 bytes)
Ferrite.facetskeleton(top, grid); # Compilation

grid = generate_grid(Hexahedron, (100, 100, 100));
@time top = ExclusiveTopology(grid);
# Master: 27.150434 seconds (64.58 M allocations: 7.195 GiB, 72.43% gc time)
#     PR:  3.799651 seconds (45 allocations: 1.895 GiB, 13.84% gc time)
@time Ferrite.facetskeleton(top, grid);
# Master: 0.143506 seconds (2 allocations: 46.234 MiB)
#     PR: 0.032492 seconds (2 allocations: 46.234 MiB)
@btime sum_skeleton_facet_neighbors($top, $grid);
# Master: 59.683 ms (0 allocations: 0 bytes)
#     PR: 21.887 ms (0 allocations: 0 bytes)
```
(PR refers to b7e2add)

### CODE REVIEW

File: src/ArrayOfVectorViews.jl
Comment (KristofferC) [inline src/ArrayOfVectorViews.jl:2]: Are these always "dense" so that you will have

```
[1:2, 3:6, 7:15, ...]
```

in that case it is unnecessary to store the second number for all of these?

Comment (KnutAM) [reply src/ArrayOfVectorViews.jl:2]: Currently they are not guaranteed to be, that would require another sorting during compress, but could be a good optimization solving both the lower index storage requirement, and speeding up iteration somewhat. (If sizehint provided to the `ConstructionBuffer` is large enough, they are already "dense" after compressing)

Comment (KristofferC) [reply src/ArrayOfVectorViews.jl:2]: My guess is that it is worth it, especially for those cases where the "inner" arrays are short.

Comment (KnutAM) [reply src/ArrayOfVectorViews.jl:2]: Yes, but that feels like an optimization we can do later? 
From the outside, everything (including type parameters) will remain the same...

Comment (KristofferC) [reply src/ArrayOfVectorViews.jl:2]: Sure

Comment (KnutAM) [reply src/ArrayOfVectorViews.jl:2]: > (If sizehint provided to the ConstructionBuffer is large enough, they are already "dense" after compressing)

My comment above was wrong, that is only true if also started `add!`ed in that order, so some signficant sorted is required because normally this is not the case.

**Comment by KnutAM** [comment]
c276cb2 brings construction in benchmark further down from 
`5.527448 seconds (58 allocations: 2.107 GiB, 9.91% gc time)`
to 
`4.217167 seconds (45 allocations: 1.895 GiB, 12.49% gc time)`

File: src/ArrayOfVectorViews.jl
Comment (termi-official) [inline src/ArrayOfVectorViews.jl:1]: Can we wrap this into a module and export only the stuff which we need?

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:1]: This should go into `src/Ferrite.jl` right?

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:145]: Maybe we should add an additional layer to also dispatch on the cell type to not overestimate too much. Where do you have the numbers from?

File: src/ArrayOfVectorViews.jl
Comment (KnutAM) [reply src/ArrayOfVectorViews.jl:1]: Yes, I think that makes sense!

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:1]: Yes, left it temporarily here but should probably move it. Will do when putting into a module!

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:145]: Yes, the commented versions are from hypercubes if I recall correctly, but an overestmate for the exclusive topology. 

Not sure if it is worthwhile to make this general since it isn't super-sensitive in my benchmarks (of course it matters, but now allocations are a quite small fraction of the time). So could even be better to simplify further and hard-code in `ExclusiveTopology`, since the same sizehints won't be valid for e.g. `CoverTopology` (didn't think of this when starting to implement this)

**Comment by KnutAM** [comment]
Thanks for the comments @termi-official - fixed the include and making into submodule. 

Do you or @koehlerson have any comments on the breaking changes that topology outputs are views of a vector instead of `EntityNeighborhood`. Is this ok?

**Comment by koehlerson** [comment]
> Thanks for the comments @termi-official - fixed the include and making into submodule.
> 
> Do you or @koehlerson have any comments on the breaking changes that topology outputs are views of a vector instead of `EntityNeighborhood`. Is this ok?

No that's fine and the `EntityNeighborhood` thing can be removed. I'll go through the changes soon :)

**Comment by KnutAM** [comment]
> No that's fine and the EntityNeighborhood thing can be removed. I'll go through the changes soon :)

Great, thanks! That was my major concern with this PR.

**Comment by KristofferC** [comment]
FWIW (without having done any detailed review) I'm fine merging this, it clearly is principally a lot better than status quo and the benchmarks suggest it is also a lot better in practice. Great job @KnutAM.

**Review by koehlerson — COMMENTED** [review]
Code looks fine to me, but the docs are not synced to the code right now. 

~~- [ ] adjust `ExclusiveTopology` struct docs~~
~~-[ ]  devdocs for the collection of views~~
~~-[ ] changelog entry~~

Sorry just noticed your task list now. Where are the docs of the `ArrayOfVectorViews`? I think it could be nice to have them in the devdocs

File: src/Grid/topology.jl
Comment (KnutAM) [inline src/Grid/topology.jl:244]:
```suggestion
    cell_neighbor_ids = Int[]
```

File: src/Grid/topology.jl
Comment (KnutAM) [inline src/Grid/topology.jl:16]:
```suggestion
#TODO: To be removed
```
You mean like this, or to remove the comment? 
Currently, `EntityNeighborhood` is still used by the `vertex_star_stencils` code. 

I guess it would make sense to change the code for this as well to return an `ArrayOfVectorViews` instead of `Dict: global_vertex_nr -> Vector{EdgeIndex}`? I didn't so far to minimize the diff, and I don't know exactly the use-cases for that function.

**Comment by KnutAM** [comment]
Thanks for the review - then I'll try to wrap it up during this week :)

I also just discussed with @fredrikekre on Slack, and we were wondering when the `include_self` keyword is used. Currently, this requires allocating a new vector (still returning a view for type-stability), potential improvements could be 

* Remove the `include_self`, leave that to the user since "self" is available anyways
* Include self as the first index in the stored `ArrayOfVectorViews`, such that `views(vals[index], 2:end)` can be returned if `include_self=false`, and `vals[index]` if `include_self=true`. But this will change the behavior if direct field access is used. 

Not required to do in this PR, but any of these changes must happen before 1.0...

File: src/Grid/topology.jl
Comment (koehlerson) [reply src/Grid/topology.jl:16]: `vertex_star_stencils` is, if I remember correctly, giving you all vertices that are connected by an edge. It can be used for mixing FEM and FDM approaches as in e.g. the topology optimization, but there the FDM operator acts on the "elements", whereas you can change that to operate on the nodes (which I think is more accurate). Would be nice to remove that. Is it currently really returning an `EdgeIndex` vector?

**Comment by koehlerson** [comment]
https://github.com/Ferrite-FEM/Ferrite.jl/blob/a10097a99cb1736999cc1bf685eaa032331fd93a/src/Grid/topology.jl#L271

e.g. here, the self reference is computed and either included or not.

**Comment by termi-official** [comment]
Maxi is correct, I need it e.g. for the distributed stuff and it is a nice convenience. Note that you can make this non-allocating by adding the self-reference during construction and return 1:N for include_self=true and 1:N-1 for include_self=false.

**Comment by termi-official** [comment]
> Not required to do in this PR, but any of these changes must happen before 1.0...

Why? The topology API is experimental and we can just add a note that we might break things here in minor releases, like we do for other stuff too.

**Comment by KnutAM** [comment]
> Note that you can make this non-allocating by adding the self-reference during construction and return 1:N for include_self=true and 1:N-1 for include_self=false.

Yes, that was option 2 above right? But I'm not sure what that will break for anyone using direct field access (not calling the neighborhood), since the returned indices will then include self?

**Comment by KnutAM** [comment]
> FWIW (without having done any detailed review) I'm fine merging this, it is clearly principally a lot better than the status quo and the benchmarks suggest it is also a lot better in practice. Great job @KnutAM.

Thanks - part of the solution was yours though :)

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:124]: Posting a quick function to get some size info for later use
```julia
function analyze_top_sizes(grid)
    top = ExclusiveTopology(grid);
    for key in fieldnames(typeof(top))
        nbh = getfield(top, key)
        nbh isa Ferrite.ArrayOfVectorViews || continue
        max_size = maximum(length, nbh; init=0)
        avg_size = sum(length, nbh; init=0)/length(nbh)
        println(key)
        println("  max: ", max_size, " entities")
        println("  avg: ", avg_size, " entities")
    end
end
```

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl]: Yes, a view of a `Vector{EdgeIndex}`.

File: docs/src/devdocs/special_datastructures.md
Comment (KristofferC) [inline docs/src/devdocs/special_datastructures.md:3]: Could the obvious question of "why not use ArraysOfArrays" be answered shortly in this section as well?

**Comment by KnutAM** [comment]
This is now ready from my perspective.
Decided to revert my attempts to include self in the saved data, simply because this broke a lot of tests and I don't have time to investigate and fix all of those. 

This should also just be an optimization and as @termi-official pointed out, the topology is considered experimental. That should probably be clarified in the docs in a separate PR before 1.0, since IMO the goal should be that 1.0 is stable unless otherwise clarly stated.

**Comment by termi-official** [comment]
> This should also just be an optimization and as @termi-official pointed out, the topology is considered experimental. That should probably be clarified in the docs in a separate PR before 1.0, since IMO the goal should be that 1.0 is stable unless otherwise clarly stated.

We clarified it already when we developed it. See https://ferrite-fem.github.io/Ferrite.jl/dev/topics/grid/#Topology . I remember that we had it in the doc string too at some point, but looks like this information has been lost in the development. Where else would you put it?

**Comment by KnutAM** [comment]
Aha, I hadn't read that. I'll just put a note in the docstring then. 
(I read the one on [release](https://ferrite-fem.github.io/Ferrite.jl/stable/reference/grid/#Topology) which gave me the impression that only ´getneighborhood` was not stable API).

**Comment by KnutAM** [comment]
When rechecking the timings, I saw some worse performance now. 
So took @termi-official's suggestion and optimized the sizehints

<details> <summary> Script </summary> 

```julia
function analyze_top_sizes(grid; show_avg=false)
    top = ExclusiveTopology(grid);
    for key in fieldnames(typeof(top))
        nbh = getfield(top, key)
        nbh isa Ferrite.ArrayOfVectorViews || continue
        max_size = maximum(length, nbh; init=0)
        avg_size = sum(length, nbh; init=0)/length(nbh)
        if show_avg
            println(key)
            println("  max: ", max_size, " entities")
            println("  avg: ", avg_size, " entities")
        else
            println(key, ": ", max_size)
        end
    end
end

function print_sizes(n=10, celltypes=[Line, Triangle, Quadrilateral, Tetrahedron, Hexahedron])
   for CT in celltypes
       grid = generate_grid(CT, ntuple(_->n, Ferrite.getrefdim(CT)))
       println(CT)
       analyze_top_sizes(grid)
       println()
   end
end
```
</details> 

Max number of entities for different generated grids

| field                   | Line | Triangle | Quadrilateral | Tetrahedron | Hexahedron |
| ----------------------- | ---- | -------- | ------------- | ----------- | ---------- |
| vertex_to_cell          | 2    | 6        | 4             | 24          | 8          |
| cell_neighbor           | 2    | 12       | 8             | 70          | 26         |
| face_face_neighbor      | 0    | 0        | 0             | 1           | 1          |
| edge_edge_neighbor      | 0    | 1        | 1             | 3           | 1          |
| vertex_vertex_neighbor  | 1    | 3        | 1             | 13          | 1          |

**Comment by KnutAM** [comment]
#988 should be fixed after merging this PR to avoid annoying conflicts.

**Comment by termi-official** [comment]
Resolves #617 ?

File: src/CollectionsOfViews.jl
Comment (termi-official) [inline src/CollectionsOfViews.jl:130]: Can we add some comment here why everything is set to zero length?

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:37]: Something in this direction maybe? Not sure how to be clear and concise here.
```suggestion
    The implementation only works with conforming grids, i.e. grids without "hanging nodes". Non-conforming grids will give unexpected results.
```

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:38]: Just for me, this is mainly a matter of not having test coverage right now?

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:126]: Can you add some commen on why this is optimal for hex? I am not seeing it directly, sorry.

File: src/Grid/topology.jl
Comment (termi-official) [inline src/Grid/topology.jl:156]: Since this is a corner case: It is expected that 1D elements dispatch on `nfaces` then?

File: src/CollectionsOfViews.jl
Comment (KnutAM) [reply src/CollectionsOfViews.jl:130]: Does this comment explain it well enough? 
```suggestion
    # Since user-code in "push_at_index!" have access to `b`, setting dimensions to zero here
    # allows GC:ing the data in `b` even in cases when the compiler cannot guarantee that 
    # it is unreachable.
    resize!(b.data, 0); sizehint!(b.data, 0)
    isa(b.indices, Vector) && (resize!(b.indices, 0); sizehint!(b.indices, 0))
```

File: test/test_grid_dofhandler_vtk.jl
Comment (termi-official) [inline test/test_grid_dofhandler_vtk.jl:356]: So the type instability is not fixed here?
```suggestion
    @test topology.edge_edge_neighbor[5,3] == Ferrite.EdgeIndex[]
    @test topology.edge_edge_neighbor[5,4] == Ferrite.EdgeIndex[]
```

(also below)

**Review by termi-official — COMMENTED** [review]
Great improvement on the performance and I think the data structure will also help in the long run. I left a few comments.

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:126]: I don't think it is easy to see due to the exclusive nature, I could add a ref to https://github.com/Ferrite-FEM/Ferrite.jl/pull/974#issuecomment-2182897133 if you think that helps?

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:38]: For embedded cells I think there will be correctness issues, since the algorithms are not written for that in mind, AFAIU from the discussion in #942 and related issue.

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:156]: This is covered by `reference_faces(::Type{RefLine}) = ()` (which gets called by `faces` getting called by `nfaces`). So it is not required to explicitly dispatch `nfaces` for a new 1D element.

File: test/test_grid_dofhandler_vtk.jl
Comment (KnutAM) [reply test/test_grid_dofhandler_vtk.jl:356]: I was not aware of this being due to type-instabilities, there should be no such for this test for sure! 
But this is indeed a strange test at least now, since e.g. 
```julia
julia> String[] == Int[]
true
```
I guess we should rather test
```suggestion
    @test isempty(topology.edge_edge_neighbor[5,3])
    @test isempty(topology.edge_edge_neighbor[5,4])
```

File: src/Grid/topology.jl
Comment (termi-official) [reply src/Grid/topology.jl:126]: Ohh, you mean for the grids generate with `generate_grid` optimal? 

Have you by any chance tried grids coming from external tools like gmsh then?

File: test/test_grid_dofhandler_vtk.jl
Comment (KnutAM) [inline test/test_grid_dofhandler_vtk.jl:359]:
```suggestion
    @test isempty(topology.edge_edge_neighbor[6,2])
    @test isempty(topology.edge_edge_neighbor[6,3])
```

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:126]: No, I found that to be out of scope for this PR since this is just an optimization. Of course for non-structured grids, there will be cases when the numbers are higher for some entities, but we probably don't won't these sizehints for the worst case as that will over-allocate. From what I've seen during development, the key is to have most connections below the sizehint. 

From my perspective, if someone wants to do a proper investigation into this, that can be done at any point, since changing these sizehints only changes the performance.

File: test/test_grid_dofhandler_vtk.jl
Comment (termi-official) [reply test/test_grid_dofhandler_vtk.jl:356]: Right

File: src/Grid/topology.jl
Comment (KnutAM) [inline src/Grid/topology.jl:114]:
```suggestion
# Guess of how many neighbors depending on grid dimension and index type. 
# This could be possible to optimize further by studying connectivities of non-uniform
# grids, see https://github.com/Ferrite-FEM/Ferrite.jl/pull/974#discussion_r1660838649
```

**Comment by KnutAM** [comment]
> I left a few comments.

Thanks!

File: src/Grid/topology.jl
Comment (koehlerson) [inline src/Grid/topology.jl:57]: Just want to highlight that this is more restrictive since before the argument was `Vector{C} where C <: AbstractCell`

File: src/Grid/topology.jl
Comment (koehlerson) [inline src/Grid/topology.jl:222]:
```suggestion
    # In this case, we loop over the cells in order and all neighbors at once.
```

**Review by koehlerson — APPROVED** [review]

File: src/Grid/topology.jl
Comment (KnutAM) [reply src/Grid/topology.jl:57]: Good point - I've added this change (and motivation) in the changelog.

**Review by termi-official — APPROVED** [review]
LGTM. Can you dump the benchmark code in the benchmark folder (here or separate PR)?

**Comment by KnutAM** [comment]
> LGTM. Can you dump the benchmark code in the benchmark folder (here or separate PR)?

Thanks! I haven't looked so much at the benchmark setup, will just open an issue for now to not forget to do it later unless someone solves it before I have time.

**Pull request merged**
